lua/nvim-treesitter/query_predicates.lua

```diff
@@ -75,6 +75,42 @@ query.add_predicate("has-ancestor?", has_ancestor)
 
 query.add_predicate("has-parent?", has_ancestor)
 
+---@param match (TSNode|nil)[]
+---@param pattern string
+---@param bufnr integer
+---@param pred string[]
+---@return boolean|nil
+local function has_child(match, pattern, bufnr, pred)
+  if not valid_args(pred[1], pred, 2) then
+    return
+  end
+
+  local node = match[pred[2]]
+  local child_types = { unpack(pred, 3) }
+  if not node then
+    return true
+  end
+
+  local just_direct_child = pred[1]:find("has-child", 1, true)
+
+  for _, child in ipairs(node:named_children()) do
+    if vim.tbl_contains(child_types, child:type()) then
+      return true
+    end
+    if not just_direct_child then
+      local child_has_child = has_child({ [pred[2]] = child }, pattern, bufnr, pred)
+      if child_has_child then
+        return true
+      end
+    end
+  end
+  return false
+end
```

Comment:
This is very expensive since you're effectively doing a query within a query. If this becomes a problem, we may want to consider adding some caching.

Reply:
`bufnr` appears unused.

Reply:
Caching would be a nice idea, ideally this shouldn't be used for more than 5 or 10 nested children imo, as then it's pretty clear the capture is matching too loosely. What could be a good depth?

Bufnr is unused but luacheck complains:
```
lua/nvim-treesitter/query_predicates.lua:83:38: (W214) used variable _pattern with unused hint
lua/nvim-treesitter/query_predicates.lua:83:48: (W214) used variable _bufnr with unused hint
```

Reply:
> Caching would be a nice idea, ideally this shouldn't be used for more than 5 or 10 nested children imo, as then it's pretty clear the capture is matching too loosely. What could be a good depth?
> 
> Bufnr is unused but luacheck complains:
> 
> ```
> lua/nvim-treesitter/query_predicates.lua:83:38: (W214) used variable _pattern with unused hint
> lua/nvim-treesitter/query_predicates.lua:83:48: (W214) used variable _bufnr with unused hint
> ```

you can use `_` instead for unused var

Reply:
There'd be two _'s which becomes very unclear what they are if in the future this function were to be modified/improved on. This would also require removing the luadoc comments above for those params

Reply:
Likely irrelevant now